### PR TITLE
Fix output for "Printing Dependent Evars Line"

### LINKED ISSUE #4504: dependent evar printout in 8.5rc1 insufficient
Note: the issue was created automatically with bugzilla2github tool

Original bug ID: BZ#4504
From: @hendriktews
Reported version: 8.5
CC:   @HuStmpHrrr

### LINKED ISSUE #10399: dependent evars line empty
In Coq 8.9.1, when I feed

    Set Printing Dependent Evars Line.
    Lemma x : exists(i : nat), i = 0.
    Proof using.
      eexists.

into `coqtop -emacs`, the dependent evars line shows one open existential:

    (dependent evars: ?X6 open,)

When I do the same on the current git master (8acc04d4c33ed41ac8dfa668a026010d4592388d), I only see

    (dependent evars:)

Was that change intentional?

(Some background: The dependent evar line was introduced by Arnaud Spiwack in 2011 for Coq 8.4 on my request for prooftree, the proof tree visualization program for Proof General and Coq. For 8.4 and 8.5 the line enables prooftree to highlight commands that introduce or instantiate a particular existential variable. Since 8.6 the feature is broken, because there is no way to link the evars appearing in the goals with those in the dependent evar line.)

### LINKED ISSUE #10420: Incomplete evars info when 'Printing Dependent Evars Line' enabled
# Feature request

Please re-enable and extend the dependent evars line, such that it shows 

1. all not instantiated evars
2. all instantiated evars and for each of them the evars appearing in the instantiating term
3. the mapping between internal evar names and their printed representation
4. all the (open) evars apearing in the current goal

For the example proof script appended below, after `apply modpon.`, the dependent evar line should look similar to

    (dependent evars: ?X4 open, ?X5 using ?X11 ?X10, ?X10 using ?X11, ?X11 open
     this goal: ?X4, ?X11
     mapping: ?P : ?X4, ?Goal2 : ?X11)

Here line 1 (realizing point 1 and 2) is identical to the dependent evars line that is printed in Coq 8.4 till 8.9 and in versions before commit 6608f64f001f8f1a50b2dc41fefdf63c0b84b270. Since that commit the dependent evars line is always empty on the master branch, see also #10399.

Point 3 (realized by line 3) is a necessary enhancement because since 8.5 there is no way to connect the evars in the proofs with the identifiers appearing in the dependent evar line.

Instead of point 4 (realized by line 2) one could extract all matches for `\?[a-z0-9]+` from the goal text. But if it is not too much effort, an explicit list inside the dependent evars line seems more robust.

The example output above means that
- evars `?X4` and `?X11` are not instantiated
-`?X5` is instantiated with a term containing `?X11` and `?X10`
-`?X10` is instantiated with a term containing `?X11`
- the current goal contains `?X4` and `?X11`
- `?X4` is printed as `?P` and `?X11` as `?Goal2`

The dependent evars line should only be printed when the flag Printing Dependent Evars Line is on. It should be printed for the current goal and for the command `Show Goal 9 at <state>` from PR #10356.

**Note**  The dependent evars output appears only in coqtop (not CoqIDE).

# Background

If this feature request is granted, prooftree (the proof visualization program for Coq/Proof General) can support the following features.
- closed branches of the proof are shown in a different color, if they still depend on not yet instantiated evars
- the evar dialog shows the dependencies among evars and can highlight proof commands that introduced or instantiated a specific evar

These features were fully working for Coq 8.4 with a dependent evar line realizing points 1 and 2 that was implemented by Arnaud Spiwack in 2011. Since Coq 8.5, however, they are more or less broken, because or the missing link between different representations of the same evar.

# Example proof script

    Set Printing Dependent Evars Line.
    Lemma strange_imp_trans : 
      forall P Q R : Prop, (Q -> R) -> (P -> Q) -> (P -> Q) -> P -> R.
    Proof.
      auto.
    Qed.

    Lemma modpon : forall P Q : Prop, (P -> Q) /\ P -> Q.
    Proof.
      tauto.
    Qed.

    Section eex.
      Variables P1 P2 P3 P4 : Prop.
      Hypothesis p12 : P1 -> P2.
      Hypothesis p123 : (P1 -> P2) -> P3.
      Hypothesis p34 : P3 -> P4.

      Lemma p14 : P4.
      Proof.
        eapply strange_imp_trans. 
              apply modpon.
            intros H.
            split.
              eexact p34.
            apply p123.
            eexact H.
          intros H.
          split.
            trivial.
          apply p123.
          trivial.
        trivial.
      Qed.
    End eex.

### PR DESCRIPTION
**Pull request by jfehrle**
Commit 6608f64f001f8f1a50b2dc41fefdf63c0b84b270 by @herbelin broke the output triggered by the `Printing Dependent Evars Line` flag.  The first commit here addresses this, fixing #10399.  There may be a better/simpler solution.  I don't know the purpose of that commit, so (very conservatively) I created a new routine `Evd.evars_of_term2`.  The routines that call this new routine are only called from `print_dependent_evars`, with `seeds` coming from `pr_open_subgoals_diff`, so I think that's safe.

The second commit addresses #4504 and #10420, a fix/enhancement adding a mapping between internal evar names and user-visible evar names.

Sample output: 

```
(dependent evars: ?X4 open, ?X5 using ?X10 ?X11, ?X10 using ?X11, ?X11 open;
mapping: ?X4 : ?P, ?X7 : ?Goal, ?X8 : ?Goal0, ?X9 : ?Goal1, ?X11 : ?Goal2)
```

Fixes / closes #4504
Fixes / closes #10399
Fixes / closes #10420

- [x] Added / updated test-suite
- [x] Corresponding documentation was added / updated (including any warning and error messages added / removed / modified).
- [x] Entry added in the changelog (see https://github.com/coq/coq/tree/master/doc/changelog#unreleased-changelog for details).

### CODE REVIEW

File: engine/evd.mli [hunk 1]
```diff
@@ -509,6 +509,9 @@ val loc_of_conv_pb : evar_map -> evar_constraint -> Loc.t option
 val evars_of_term : evar_map -> econstr -> Evar.Set.t
   (** including evars in instances of evars *)
 
+val evars_of_term2 : evar_map -> econstr -> Evar.Set.t
```
Comment (SkySkimmer) [inline engine/evd.mli]: This should be of type `constr -> Evar.Set.t` (the evar map argument isn't used so don't have it)

**Comment by jfehrle** [comment]
@herbelin Is there an easy way to get the list of evars that are visible in the current goal?

File: doc/changelog/07-commands-and-options/10483-print_dependent_evars.rst [hunk 2]
```diff
@@ -0,0 +1,7 @@
+- Update output generated by "Printing Dependent Evars Line" option
```
Comment (Zimmi48) [inline doc/changelog/07-commands-and-options/10483-print_dependent_evars.rst]:
```suggestion
- Update output generated by :flag:`Printing Dependent Evars Line` flag
```

File: doc/sphinx/proof-engine/vernacular-commands.rst [hunk 3]
```diff
@@ -1031,8 +1031,9 @@ Controlling display
 
 .. flag:: Printing Dependent Evars Line
 
-   This option controls the printing of the “(dependent evars: …)” line when
-   ``-emacs`` is passed.
+   This option controls the printing of the “(dependent evars: …)” information
+   after each tactic.  The information is used by the Prooftree tool in Proof
+   General. (https://askra.de/software/prooftree)
```
Comment (Zimmi48) [inline doc/sphinx/proof-engine/vernacular-commands.rst]: Why remove the info that this is used when the `-emacs` flag is passed?

Comment (jfehrle) [reply doc/sphinx/proof-engine/vernacular-commands.rst]: The code no longer requires `-emacs` to see the variables.  Originally -emacs alone triggered the printing before there was the option.  Once you have the option the -emacs is not needed.  So just correcting the doc, not something I changed.

**Review by Zimmi48 — APPROVED** [review]

**Comment by Zimmi48** [comment]
Warning: you have pushed this branch to the upstream repository instead of your fork. Remember to delete it when this PR is merged and avoid doing this in the future.

**Comment by jfehrle** [comment]
Content moved to #10489.

**Comment by jfehrle** [comment]
> Warning: you have pushed this branch to the upstream repository instead of your fork. Remember to delete it when this PR is merged and avoid doing this in the future.

From a new misconfigured source tree.

**Pull request closed without merging**